GraphNeuralNetworks/src/samplers.jl

```diff
@@ -0,0 +1,110 @@
+"""
+    struct NeighborLoader
+
+A data structure for sampling neighbors from a graph for training Graph Neural Networks (GNNs). 
+It supports multi-layer sampling of neighbors for a batch of input nodes, useful for mini-batch training
+originally introduced in "Inductive Representation Learning on Large Graphs" paper.
+[see https://arxiv.org/abs/1706.02216]
+
+# Fields
+- `graph::GNNGraph`: The input graph.
+- `num_neighbors::Vector{Int}`: A vector specifying the number of neighbors to sample per node at each GNN layer.
+- `input_nodes::Vector{Int}`: A vector containing the starting nodes for neighbor sampling.
+- `num_layers::Int`: The number of layers for neighborhood expansion (how far to sample neighbors).
+- `batch_size::Union{Int, Nothing}`: The size of the batch. If not specified, it defaults to the number of `input_nodes`.
+
+# Usage
+```julia
+loader = NeighborLoader(graph; num_neighbors=[10, 5], input_nodes=[1, 2, 3], num_layers=2)
+```
+"""
+struct NeighborLoader
+    graph::GNNGraph             # The input GNNGraph (graph + features from GraphNeuralNetworks.jl)
+    num_neighbors::Vector{Int}  # Number of neighbors to sample per node, for each layer
+    input_nodes::Vector{Int}    # Set of input nodes (starting nodes for sampling)
+    num_layers::Int             # Number of layers for neighborhood expansion
+    batch_size::Union{Int, Nothing}  # Optional batch size, defaults to the length of input_nodes if not given
+    neighbors_cache::Dict{Int, Vector{Int}}  # Cache neighbors to avoid recomputation
+end
+
+function NeighborLoader(graph::GNNGraph; num_neighbors::Vector{Int}, input_nodes::Vector{Int}, num_layers::Int, batch_size::Union{Int, Nothing}=nothing)
```

Comment:
`input_node` could have default value = `nothing`. If not passed it should consider all nodes in the graph.